internal/defaultingress/common.go

```diff
@@ -21,49 +21,144 @@ package defaultingress
+				"more_set_headers \"X-Content-Type-Options: nosniff\";\n" +
+				"more_set_headers \"Referrer-Policy: no-referrer-when-downgrade\";",
+		}
+		return annotations
 	}
 
-	// we don't want the metrics to be exposed on /<servicename>/metrics. Thus, always return 404
-	serverSnippetValue := fmt.Sprint("location ~* \"/(appengine|flow|housekeeping|pairing|realmmanagement)/metrics\" {\n" +
-		"  deny all;\n" +
-		"  return 404;\n" +
-		"}")
-	annotations["nginx.ingress.kubernetes.io/server-snippet"] = serverSnippetValue
+	// From here on, we assume HAProxy Ingress Controller is in use
 
-	if pointy.BoolValue(cr.Spec.API.Cors, false) {
-		annotations["nginx.ingress.kubernetes.io/enable-cors"] = strconv.FormatBool(true)
+	annotations = map[string]string{
+		"haproxy.org/backend-config-snippet": getHAProxyBackendConfig(parent),
+		"haproxy.org/ssl-redirect":           strconv.FormatBool(apiSslRedirect),
+		"haproxy.org/response-set-header": "\n" +
+			"X-Frame-Options SAMEORIGIN\n" +
+			"Content-Security-Policy " + buildContentSecurityPolicy(parent) + "\n" +
+			"X-Content-Type-Options nosniff\n" +
+			"Referrer-Policy no-referrer-when-downgrade",
 	}
+	annotations = appendHAProxyCorsAnnotations(enableCors, annotations)
+
+	return annotations
+}
+
+// appendHAProxyCorsAnnotations appends the necessary HAProxy annotations
+// depending on whether CORS is enabled or not.
+func appendHAProxyCorsAnnotations(enableCors bool, annotations map[string]string) (ret map[string]string) {
+	if !enableCors {
+		annotations["haproxy.org/cors-enable"] = "false"
+		return annotations
+	}
+
+	annotations["haproxy.org/cors-enable"] = "true"
+	annotations["haproxy.org/cors-allow-origin"] = "*"
+	annotations["haproxy.org/cors-allow-methods"] = "GET, POST, OPTIONS, PUT, DELETE"
+	annotations["haproxy.org/cors-allow-headers"] = "DNT,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Range,Authorization"
+	annotations["haproxy.org/cors-expose-headers"] = "Content-Length,Content-Range"
 
 	return annotations
 }
 
-// TODO handle with kubebuilder defaults
-func getIngressClassName(cr *ingressv1alpha1.AstarteDefaultIngress) *string {
+// getHAProxyBackendConfig returns the backend config snippet for HAProxy Ingresses
+// Performs path rewriting only for Astarte API paths, leaving others (like dashboard) untouched.
+func getHAProxyBackendConfig(cr *apiv2alpha1.Astarte) string {
+	return fmt.Sprintf(`http-request replace-path /(appengine|pairing|housekeeping|realmmanagement)/(.*) /\2 if { hdr(host) -i %s }`, cr.Spec.API.Host)
+}
+
+// getIngressClassName returns the Ingress Class Name to use
+// ADI annotation is used to select the proper default ingressClass
+func getIngressClassName(cr *ingressv2alpha1.AstarteDefaultIngress) string {
 	if cr.Spec.IngressClass == "" {
-		return pointy.String("nginx")
+		if useHAProxyIngressController(cr) {
+			return "haproxy"
+		}
+		return "nginx"
+	}
+	return cr.Spec.IngressClass
+}
```

Comment:
Just wondering if it belongs here. Should it be part of the defaulting webhook to inject default values when no user input is provided?